File: backend/src/models/GatewayCustomer.ts
import mongoose, { Schema, Document } from 'mongoose';
import { IUser } from './User';

export interface IGatewayCustomer extends Document {
  ownerId: IUser['_id'];
  name: string;
  email?: string;
  phoneNumber: string;
  totalTransactions: number;
  totalAmount: number;
  lastTransactionDate?: Date;
  lastTransactionStatus?: 'PENDING' | 'SUCCESS' | 'FAILED';
  notes?: string;
  tags?: string[];
  createdAt: Date;
  updatedAt: Date;
}

const GatewayCustomerSchema: Schema = new Schema({
  ownerId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  name: { type: String, required: true },
  email: { type: String },
  phoneNumber: { type: String, required: true },
  totalTransactions: { type: Number, default: 0 },
  totalAmount: { type: Number, default: 0 },
  lastTransactionDate: { type: Date },
  lastTransactionStatus: { type: String, enum: ['PENDING', 'SUCCESS', 'FAILED'] },
  notes: { type: String },
  tags: [{ type: String }],
}, { timestamps: true });

GatewayCustomerSchema.index({ ownerId: 1 });
GatewayCustomerSchema.index({ phoneNumber: 1 });
GatewayCustomerSchema.index({ email: 1 });

export default mongoose.model<IGatewayCustomer>('GatewayCustomer', GatewayCustomerSchema);

File: backend/app/models/gateway_customer.py
from datetime import datetime, timezone
from typing import Optional, Literal
from beanie import PydanticObjectId
from pydantic import Field, field_validator # Import field_validator
import re # Import re module for regex
from app.models.base import BaseDocument

class GatewayCustomer(BaseDocument):
    owner_id: PydanticObjectId = Field(alias="ownerId")
    name: str
    email: Optional[str] = None
    phone_number: str = Field(alias="phoneNumber")
    total_transactions: int = Field(default=0, alias="totalTransactions")
    total_amount: float = Field(default=0, alias="totalAmount")
    last_transaction_date: Optional[datetime] = Field(
        default=None, alias="lastTransactionDate"
    )
    last_transaction_status: Optional[Literal["PENDING", "SUCCESS", "FAILED"]] = Field(
        default=None, alias="lastTransactionStatus"
    )
    notes: Optional[str] = None
    tags: Optional[list[str]] = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "gatewaycustomers"
        indexes = [
            "owner_id",
            "phone_number",
            "email",
        ]

    @field_validator("phone_number")
    @classmethod
    def validate_phone(cls, v: str) -> str:
        if not re.match(r"^254(1|7)\d{8}$", v):
            raise ValueError("Phone number must be in format 254XXXXXXXX")
        return v
